File: src/node.rs
use core::str;
use std::{
    borrow::{Borrow, BorrowMut},
    io::Error,
    ops::Deref,
};

use crate::{
    data_access_layer::DataAccessLayer,
    page::{PageNum, PAGE_NUM_SIZE},
};

#[derive(Clone)]
pub struct Item {
    pub key: Vec<u8>,
    pub value: Vec<u8>,
}

pub struct Node<'dal> {
    pub dal: &'dal mut DataAccessLayer,
    pub page_num: PageNum,
    pub items: Vec<Item>,
    pub children: Vec<PageNum>,
}

pub const NODE_HEADER_SIZE: usize = 3;

impl Item {
    pub fn new(key: Vec<u8>, value: Vec<u8>) -> Self {
        Self { key, value }
    }

    #[inline]
    pub fn size(&self) -> usize {
        self.key.len() + self.value.len() + PAGE_NUM_SIZE
    }
}

enum SplitMethod<'dal> {
    LEAF(Node<'dal>),
    ELSE(Node<'dal>),
}

impl<'dal> Node<'dal> {
    pub fn new(dal: &'dal mut DataAccessLayer) -> Self {
        Self {
            dal,
            page_num: 0,
            items: Vec::new(),
            children: Vec::new(),
        }
    }

    #[inline]
    pub fn item_size(&self, index: usize) -> usize {
        self.items[index].size()
    }

    pub fn node_size(&self) -> usize {
        self.items
            .iter()
            .fold(NODE_HEADER_SIZE, |size, item| size + item.size())
            + PAGE_NUM_SIZE
    }

    #[inline]
    pub fn is_leaf(&self) -> bool {
        self.children.len() == 0
    }

    pub fn add_item(&mut self, item: Item, index: usize) -> usize {
        if index == self.items.len() {
            self.items.push(item);
            return index;
        }

        self.items.insert(index, item);
        index
    }

    pub fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), Error> {
        let mut left = 0;
        let mut right = 0;
        let is_leaf = self.is_leaf();

        buf[left] = if self.is_leaf() { 0u8 } else { 1u8 };
        left += 1;

        buf[left..left + 2].copy_from_slice(&(self.items.len() as u16).to_le_bytes());
        left += 2;

        let mut child_index = 0;
        self.items.iter().for_each(|item| {
            if !is_leaf {
                buf[left..left + PAGE_NUM_SIZE].copy_from_slice(&(self.children[0]).to_le_bytes());
                left += PAGE_NUM_SIZE;
            }

            buf[left..left + 2].copy_from_slice(
                &((right - item.key.len() - item.value.len() - 2) as u16).to_le_bytes(),
            );
            left += 2;

            right -= item.value.len();
            buf[right..].copy_from_slice(&item.value);

            right -= 1;
            buf[right] = item.value.len() as u8;

            right -= item.key.len();
            buf[right..].copy_from_slice(&item.key);

            right -= 1;
            buf[right] = item.key.len() as u8;

            child_index += 1;
        });

        if !is_leaf {
            buf[left..left + PAGE_NUM_SIZE].copy_from_slice(
                &(self
                    .children
                    .last()
                    .expect("serialize error: no last child node"))
                .to_le_bytes(),
            );
        }

        Ok(())
    }

    pub fn deserialize(&mut self, buf: &Vec<u8>) {
        let mut left = 0;
        let is_leaf = buf[0] as u16 == 1;

        left += 3;
        for i in 0..(u16::from_le_bytes(
            buf[1..3]
                .try_into()
                .expect("deserialize error: cannot read the items count"),
        )) {
            if !is_leaf {
                self.children.push(PageNum::from_le_bytes(
                    buf[left..left + PAGE_NUM_SIZE]
                        .try_into()
                        .expect(format!("deserialize error: cannot read the child {}", i).as_str()),
                ));
                left += PAGE_NUM_SIZE;
            }

            let mut offset = u16::from_le_bytes(
                buf[left..left + 2]
                    .try_into()
                    .expect("deserialize error: cannot read the offset"),
            ) as usize;

            left += 2;

            let key_len = buf[offset] as usize;
            offset += 1;

            let key = (buf[offset..offset + key_len])
                .iter()
                .cloned()
                .collect::<Vec<u8>>();
            offset += key_len;

            let value_len = buf[offset] as usize;
            offset += 1;

            let value = (buf[offset..offset + value_len])
                .iter()
                .cloned()
                .collect::<Vec<u8>>();

            self.items.push(Item::new(key, value));
        }

        if !is_leaf {
            self.children.push(PageNum::from_le_bytes(
                buf[left..left + PAGE_NUM_SIZE]
                    .try_into()
                    .expect("deserialize error: cannot read the last child"),
            ));
        }
    }

    fn __find_key_in_node(&self, key: &Vec<u8>) -> Result<(bool, usize), Error> {
        for (index, item) in self.items.iter().enumerate() {
            match item.key.cmp(key) {
                std::cmp::Ordering::Less => continue,
                std::cmp::Ordering::Equal => return Ok((true, index)),
                std::cmp::Ordering::Greater => return Ok((false, index)),
            }
        }

        Ok((false, self.items.len()))
    }

    /// TODO: &'dal Node<'dal>
    fn __find_key_helper(
        node: Node<'dal>,
        key: &'dal Vec<u8>,
    ) -> Result<(usize, Option<Node<'dal>>), Error> {
        let (was_found, index) = node.__find_key_in_node(key)?;

        if was_found {
            return Ok((index, Some(node)));
        }

        if node.is_leaf() {
            return Ok((0, None));
        }

        return Self::__find_key_helper(node.dal.get_node(node.children[index])?, key);
    }

    /// TODO: &self
    pub fn find(self, key: &'dal Vec<u8>) -> Result<(usize, Option<Node>), Error> {
        Self::__find_key_helper(self, key)
    }

    #[inline]
    pub fn is_over_populated(&self) -> bool {
        self.dal.is_over_populated(self)
    }

    #[inline]
    pub fn is_under_populated(&self) -> bool {
        self.dal.is_under_populated(self)
    }

    fn __split_new_node(&mut self, node_to_split: &mut Node, split_index: usize) -> SplitMethod {
        if node_to_split.is_leaf() {
            return SplitMethod::LEAF(
                self.dal
                    .new_node(node_to_split.items[split_index + 1..].to_vec(), vec![]),
            );
        } else {
            return SplitMethod::ELSE(self.dal.new_node(
                node_to_split.items[split_index + 1..].to_vec(),
                node_to_split.children[split_index + 1..].to_vec(),
            ));
        }
    }

    fn __handle_split_new_node(&mut self, node_to_split: &mut Node, index: usize) -> Node {
        let split_index = node_to_split
            .dal
            .get_split_index(node_to_split)
            .expect("node split: cannot get the split index");

        match self.__split_new_node(node_to_split, split_index) {
            SplitMethod::LEAF(node) => {
                node_to_split.items = node_to_split.items[..split_index].to_vec();
                node
            }
            SplitMethod::ELSE(node) => {
                node_to_split.items = node_to_split.items[..split_index].to_vec();
                node_to_split.children = node_to_split.children[..split_index + 1].to_vec();
                node
            }
        }
    }
}

File: core/node.cpp
#include "node.h"
#include "errors.hpp"
#include "page.h"
#include "data_access_layer.h"
#include <cstdarg>
#include <cstddef>
#include <numeric>
#include <span>
#include <stdexcept>
#include "utils.h"

namespace toocal::core::node
{
  [[nodiscard]] auto Node::is_leaf() const noexcept -> bool { return this->children.empty(); }

  [[nodiscard]] auto Item::size() const noexcept -> uint32_t
  {
    return this->key.size() + this->value.size();
  }

  [[nodiscard]] auto Node::item_size(uint32_t index) const noexcept -> uint32_t
  {
    try
      {
        const auto& [key, value] = this->items[index];
        return key.size() + value.size() + sizeof(Page::page_num);
      }
    catch (const std::out_of_range& e)
      {
        fatal(fmt::format("cannot get the item size of {} because out of range", index));
      }
  }

  [[nodiscard]] auto Node::size() const noexcept -> uint32_t
  {
    return std::accumulate(
      items.begin(),
      items.end(),
      static_cast<uint32_t>(Node::HEADER_SIZE + sizeof(page::Page_num)),
      [&](const uint32_t size, const auto& item) { return size + item.size(); });
  }

  [[nodiscard]] auto Node::is_last(const uint32_t index, const Node& parent_node) const noexcept
    -> bool
  {
    return index == parent_node.items.size();
  }

  [[nodiscard]] auto Node::is_first(const uint32_t index) const noexcept -> bool
  {
    return index == 0;
  }

  [[nodiscard]] auto Node::find_key_helper(
    const Node&                 node,
    const std::vector<uint8_t>& key,
    const bool                  exact,
    std::deque<uint32_t>&       ancestors_indexes) noexcept
    -> tl::expected<std::tuple<int, tl::optional<Node>>, Error>
  {
    const auto&& [was_found, index] = node.find_key_in_node(key);
    if (was_found)
      return std::make_tuple(index, node);

    if (node.is_leaf())
      {
        if (exact)
          return std::make_tuple(-1, tl::nullopt);
        return std::make_tuple(index, node);
      }

    ancestors_indexes.push_back(index);
    const auto next_child = node.dal->get_node(node.children[index]).map_error([](auto&& error) {
      error.append("get_node error in find_key_helper");
      return error;
    });

    if (next_child.has_value())
      return find_key_helper(next_child.value(), key, exact, ancestors_indexes);
    return tl::unexpected(next_child.error());
  }

  [[nodiscard]] auto Node::find_key_in_node(const std::vector<uint8_t>& key) const noexcept
    -> std::tuple<bool, uint32_t>
  {
    for (uint32_t index = 0; const auto& [existing_item_key, _] : this->items)
      {
        const auto compare_result = utils::Safecmp::bytescmp(existing_item_key, key);
        if (compare_result == 0)
          return {true, index};

        /* The key is bigger than the previous item, so it doesn't exist in the
         * node, but may exist in child nodes. */
        if (compare_result > 0)
          return {false, index};
        ++index;
      }

    /* The key is bigger than the previous item, so it doesn't exist in the
     * node, but may exist in child nodes. */
    return {false, this->items.size()};
  }

  [[nodiscard]] auto Node::find_key(const std::vector<uint8_t>& key, const bool exact) const noexcept
    -> tl::expected<std::tuple<int, tl::optional<Node>, std::deque<uint32_t>>, Error>
  {
    auto ancestors_indexes = std::deque<uint32_t>{/* index of root */ 0};

    return find_key_helper(*this, key, exact, ancestors_indexes).map([&](const auto& result) {
      const auto& [index, node] = result;
      return std::make_tuple(index, node, ancestors_indexes);
    });
  }

  auto Node::add_item(const Item& item, const uint32_t insertion_index) noexcept -> uint32_t
  {
    if (this->items.size() == insertion_index)
      {
        this->items.push_back(item);
        return insertion_index;
      }
    this->items.insert(this->items.begin() + insertion_index, item);
    return insertion_index;
  }

  [[nodiscard]] auto Node::is_over_populated() const noexcept -> bool
  {
    return this->dal->is_over_populated(*this);
  }

  [[nodiscard]] auto Node::is_under_populated() const noexcept -> bool
  {
    return this->dal->is_under_populated(*this);
  }

  static auto _split_when_node_is_leaf(
    Data_access_layer* dal, Node& node_to_split, const uint32_t split_index, Node& new_node) noexcept
    -> void
  {
    auto node = dal->new_node(
      std::deque<Item>{node_to_split.items.begin() + split_index + 1, node_to_split.items.end()},
      std::deque<page::Page_num>{});

    dal->write_node(node)
      .map([&](const auto&& _) {
        new_node = std::move(node);
        node_to_split.items =
          std::deque<Item>{node_to_split.items.begin(), node_to_split.items.begin() + split_index};

        return nullptr;
      })
      .map_error([&](auto&& error) {
        error.append("write_node error in Node::split");
        return error.panic();
      });
  }

  static auto _split_when_node_is_not_leaf(
    Data_access_layer* dal, Node& node_to_split, const uint32_t split_index, Node& new_node) noexcept
    -> void
  {
    auto node = dal->new_node(
      std::deque<Item>{node_to_split.items.begin() + split_index + 1, node_to_split.items.end()},
      std::deque<page::Page_num>{
        node_to_split.children.begin() + split_index + 1, node_to_split.children.end()});

    dal->write_node(node)
      .map([&](const auto&& _) {
        new_node = std::move(node);
        node_to_split.items =
          std::deque<Item>{node_to_split.items.begin(), node_to_split.items.begin() + split_index};

        node_to_split.children = std::deque<page::Page_num>{
          node_to_split.children.begin(), node_to_split.children.begin() + split_index + 1};

        return nullptr;
      })
      .map_error([&](auto&& error) {
        error.append("write_node error in Node::split");
        return error.panic();
      });
  }

  auto Node::split(Node& node_to_split, const uint32_t node_to_split_index) noexcept -> void
  {
    /* the first index where min amount of bytes to populate a page is archived.
     * then add 1 so it will be split one index after. */
    const auto split_index = node_to_split.dal->get_split_index(node_to_split);
    const auto middle_item = node_to_split.items[split_index];

    Node new_node;
    if (node_to_split.is_leaf())
      _split_when_node_is_leaf(this->dal, node_to_split, split_index, new_node);
    else
      _split_when_node_is_not_leaf(this->dal, node_to_split, split_index, new_node);

    this->add_item(middle_item, node_to_split_index);

    /* if middle of list, then move items forward. */
    if (this->children.size() == node_to_split_index + 1)
      this->children.push_back(new_node.page_num);
    else
      this->children.insert(this->children.begin() + node_to_split_index + 1, new_node.page_num);

    this->dal->write_node(*this)
      .and_then([&](const auto&& _) { return this->dal->write_node(node_to_split); })
      .map_error([](auto&& error) {
        error.append("write_node error in Node::split");
        return error.panic();
      });
  }

  auto Node::remove_item_from_leaf(int32_t index) noexcept -> void
  {
    this->items.erase(this->items.begin() + index);
    this->dal->write_node(*this).map_error([&](auto&& error) {
      error.append("write_node error in Node::remove_items_from_leaf");
      return error.panic();
    });
  }

  [[nodiscard]] auto Node::remove_item_from_internal(int32_t index) noexcept
    -> tl::expected<std::deque<int32_t>, Error>
  {
    auto affected_nodes = std::deque<int32_t>{index};
    return this->dal->get_node(this->children[index])
      .map([&](auto&& node) {
        /* starting from its left child, descend to the rightmost descendant. */
        while (!node.is_leaf())
          {
            const auto traversing_index = this->children.size() - 1;
            node.dal->get_node(node.children[traversing_index])
              .map([&](const auto&& _) {
                affected_nodes.push_back(traversing_index);
                return nullptr;
              })
              .map_error([](auto&& error) {
                error.append("get_node error in Node::remote_item_from_internal");
                return error.panic();
              });
          }

        /* replace the item that should be removed with
         * the item before inorder which we just found. */
        node.items.pop_back();

        return this->dal->write_node(*this)
          .map([&](const auto&& _) { return this->dal->write_node(node); })
          .map_error([](auto&& error) {
            error.append("write_node error in Node::remote_item_from_internal");
            return error.panic();
          });
      })
      .map([&](const auto&& _) { return affected_nodes; });
  }

  auto Node::rotate_right(Node& anode, Node& pnode, Node& bnode, int32_t bnode_index) noexcept
    -> void
  {
    /* get last item and remove it */
    const auto anode_item = anode.items.back();
    anode.items.pop_back();

    /* get item from parent node and assign the aNodeItem item instead. */
    const auto pnode_item_index = (this->is_first(bnode_index)) ? 0 : bnode_index - 1;
    const auto pnode_item = pnode.items[pnode_item_index];

    pnode.items[pnode_item_index] = anode_item;

    /* assign parent item to b and make it first */
    bnode.items.push_front(pnode_item);

    /* if it's an inner leaf then move children as well. */
    if (!anode.is_leaf())
      {
        const auto child_to_shift = anode.children.back();
        anode.children.pop_back();
        bnode.children.push_front(child_to_shift);
      }

    return;
  }

  auto Node::rotate_left(Node& anode, Node& pnode, Node& bnode, int32_t bnode_index) noexcept
    -> void
  {
    /* get first item and remove it */
    const auto bnode_item = bnode.items[0];
    bnode.items.pop_front();

    /* get item from parent node and assign the bNodeItem item instead */
    const auto pnode_item_index =
      (this->is_last(bnode_index, pnode)) ? (pnode.items.size() - 1) : bnode_index;

    const auto pnode_item = pnode.items[pnode_item_index];
    pnode.items[pnode_item_index] = bnode_item;

    /* assign parent item to a and make it last */
    anode.items.push_back(pnode_item);

    /* if it's an inner leaf then move children as well. */
    if (!bnode.is_leaf())
      {
        const auto child_to_shift = bnode.children[0];
        bnode.children.pop_front();
        anode.children.push_back(child_to_shift);
      }

    return;
  }

  [[nodiscard]] auto Node::merge(Node& bnode, int32_t bnode_index) noexcept
    -> tl::expected<std::nullptr_t, Error>
  {
    return this->dal->get_node(this->children[bnode_index - 1]).and_then([&](auto&& node) {
      /* take the item from the parent, remove it and add it to the unbalanced node */
      const auto pnode_item = this->items[bnode_index - 1];
      this->items.erase(this->items.begin() + bnode_index - 1);
      node.items.push_back(pnode_item);

      if (!node.is_leaf())
        node.children.insert(node.children.end(), bnode.children.begin(), bnode.children.end());

      return this->dal->write_node(node)
        .and_then([&](const auto&& _) { return this->dal->write_node(*this); })
        .map([&](const auto&& _) {
          this->dal->delete_node(bnode.page_num);
          return nullptr;
        });
    });
  }

  [[nodiscard]] auto
    Node::rebalance_remove(Node& unbalanced_node, int32_t unbalanced_node_index) noexcept
    -> tl::expected<std::nullptr_t, Error>
  {
    auto pnode = *this;
    /* right rotate */
    if (unbalanced_node_index != 0)
      {
        this->dal->get_node(pnode.children[unbalanced_node_index - 1])
          .and_then([&](auto&& left_node) -> tl::expected<std::nullptr_t, Error> {
            if (left_node.can_spare_an_element())
              {
                rotate_right(left_node, pnode, unbalanced_node, unbalanced_node_index);
                return this->dal->write_node(left_node)
                  .and_then([&](const auto&& _) { return this->dal->write_node(pnode); })
                  .and_then([&](const auto&& _) { return this->dal->write_node(unbalanced_node); });
              }

            return nullptr;
          });
      }

    /* left balance */
    if (unbalanced_node_index != pnode.children.size() - 1)
      {
        this->dal->get_node(pnode.children[unbalanced_node_index + 1])
          .and_then([&](auto&& right_node) -> tl::expected<std::nullptr_t, Error> {
            if (right_node.can_spare_an_element())
              {
                rotate_left(unbalanced_node, pnode, right_node, unbalanced_node_index);
                return this->dal->write_node(unbalanced_node)
                  .and_then([&](const auto&& _) { return this->dal->write_node(pnode); })
                  .and_then([&](const auto&& _) { return this->dal->write_node(right_node); });
              }

            return nullptr;
          });
      }

    /* The merge function merges a given node with its node to the right.
     * So by default, we merge an unbalanced node with its right sibling.
     * In the case where the unbalanced node is the leftmost,
     * we have to replace the merge parameters, so the unbalanced node right sibling,
     * will be merged into the unbalanced node. */
    if (unbalanced_node_index == 0)
      return this->dal->get_node(this->children[unbalanced_node_index + 1])
        .and_then([&](auto&& node) { return pnode.merge(node, unbalanced_node_index + 1); });

    return pnode.merge(unbalanced_node, unbalanced_node_index);
  }

  [[nodiscard]] auto Node::can_spare_an_element() const noexcept -> bool
  {
    if (this->dal->get_split_index(*this) == -1)
      return false;
    return true;
  }
} // namespace toocal::core::node
